generator/testdata/google/example/library/v1/library_service.gapic.cc.baseline

```diff
@@ -9,84 +11,192 @@
 gax::StatusOr<::google::example::library::v1::Book>
 LibraryService::CreateBook(
```

Comment:
I was going to comment on this in the design doc, but ... by making the retry loop part of the `Service` class (as opposed to a Decorator) you stop everybody from putting a decorator around the retry loop.  And I know we want to put at least the OpenCensus decorator around the retry loop so we know that all the retries are part of the same span.

Reply:
@michaelbausor 
Is this an indication that we should revist ClientStub or provide a similar degree of user customization?

Reply:
I see two ways to support implementing retries as a decorator: (1) have retries be implemented as a decorator on the GAPIC Stub class, or (2) provide an additional abstract layer that can be decorated, like the ClientStub concept.

The problems I see with (1):
 - We are committed to using the grpc types (grpc::ClientContext, grpc::Status) in our RetryPolicy, and any other policies that are implemented as decorators. This is likely to be fine in terms of functionality, but means that these types are now a more significant part of our public interface.
 - How exactly will users apply decorators in this model? Right now they can write their own Stub decorators, apply them to a Stub, and pass it into the Client, without the client being aware of it. The model used by google-cloud-cpp Storage is to provide a cosntructor that takes a RawClient (equivalent of our Stub) and specifies NoDecorators. So users who make use of this need to specify their own retries, and in the case that we implement anything else via decorators (logging, x-goog-api-client metadata) would need to also specify those.

The problems I see with (2):
 - Providing two places that allow decoration, with different interfaces, seems overly complex, and would be great to avoid if at all possible.
 - We would be committing to hiding the grpc stub surface. I think this is a major commitment that we rejected during the discussions on the gapic stub. I don't want to say no out of hand, but I think we would need to have a really compelling case to do it, and a clear plan for handling async and streaming methods.

My suggested solution would be to provide first-class support for OpenCensus in the form of a Policy object. This is less flexible that a decorator-based solution (it doesn't help with other examples of decorators that need to be above the retries) but IMO is much less complex. Are there other problems with this approach? And, are there other (known or just suspected) cases of needing to decorate the Stub above the level of retries?

Reply:
Also - apologies for the essay :-/

Reply:
Another consideration - does implementing retries as a decorator make it impossible to allow method-level configuration of retries? e.g. if we wanted to provide a method override like:

```c++
StatusOr<EchoResponse> Echo(EchoRequest const& request);
StatusOr<EchoResponse> Echo(EchoRequest const& request, Policies&&...);
```
where one of the valid Policy objects would be a RetryPolicy? It seems like that would be impossible if retries are implemented as part of an opaque Stub stack.

Reply:
@michaelbausor 

> My suggested solution would be to provide first-class support for OpenCensus in the form of a Policy object. This is less flexible that a decorator-based solution (it doesn't help with other examples of decorators that need to be above the retries) but IMO is much less complex.

> Are there other problems with this approach?

None that are immediate, except that the decorators play nicer (I think) with the following ideas.

> And, are there other (known or just suspected) cases of needing to decorate the Stub above the level of retries?

A couple:

- We want to implement a decorator that suppresses duplicate requests. The typical use-case is "read-the-value-for-this-key", you want to have ~1 of those in flight, because all the responses are going to be approximately the same, specially when the key does not exist :grin:  (Note that this is not caching, it is simply "suppress a duplicate request and reuse the response").  In production, at scale, this can improve recovery from hot-spotting into a single database row or object.  You need to decorate all functions because you want to prevent RAW hazards.

- We want to implement a decorator that immediately fails a request when we know the server is overloaded (how we "know" that is a matter for a large design doc, but they do exist inside Google).  When that happens there is a percentage of the requests that is better to *not* send and fail, rather than send-and-retry-and-create-event-more-load in the server-side.

This may give you other ideas:

https://github.com/googleapis/google-cloud-cpp/milestone/8

Reply:
@coryan I'm thinking through this issue at the moment, particularly in relation to our proposed gax::CallContext class. One question - why did google-cloud-cpp choose to use the decorator model for Storage, but not for Bigtable? Is it just historical reasons, and you would use the decorator model for both if starting again today?

Reply:
Purely historical reasons, we wrote the Cloud Bigtable C++ client first, and we did not know about a number of requirements at the time (logging, OpenCensus integration, suppress duplicate requests).

I would write Bigtable with decorators today. In fact, I am trying to figure out a way to introduce them into the code already.